File: typescript/src/approval.ts
import type { Transport } from "./transport.js";
import type {
  ApprovalStatus,
  DecideApprovalResult,
  WaitForApprovalOptions,
} from "./types.js";
import {
  CanopyApprovalTimeoutError,
  CanopyChatApprovalDisabledError,
} from "./errors.js";

interface ApprovalStatusResponse {
  status: "pending" | "approved" | "denied" | "expired";
  decided_at: string | null;
  expires_at: string;
  transaction_id: string;
  x_payment_header: string | null;
  mpp_payment_header?: string | null;
}

export async function getApprovalStatus(
  transport: Transport,
  approvalId: string,
): Promise<ApprovalStatus> {
  const { body } = await transport.request<ApprovalStatusResponse>({
    method: "GET",
    path: `/api/approvals/${approvalId}/status`,
    expectStatuses: [200],
  });
  return {
    status: body.status,
    decidedAt: body.decided_at,
    expiresAt: body.expires_at,
    transactionId: body.transaction_id,
    xPaymentHeader: body.x_payment_header ?? null,
    mppPaymentHeader: body.mpp_payment_header ?? null,
  };
}

/**
 * Polls the approval's status until it leaves `pending`, timing out after
 * `timeoutMs` (default 5 minutes). Polling interval is 2s by default.
 */
export async function waitForApproval(
  transport: Transport,
  approvalId: string,
  opts: WaitForApprovalOptions = {},
): Promise<ApprovalStatus> {
  const timeoutMs = opts.timeoutMs ?? 5 * 60 * 1000;
  const pollIntervalMs = opts.pollIntervalMs ?? 2000;
  const deadline = Date.now() + timeoutMs;

  while (true) {
    const status = await getApprovalStatus(transport, approvalId);
    if (status.status !== "pending") return status;
    if (Date.now() >= deadline) {
      throw new CanopyApprovalTimeoutError(approvalId, timeoutMs);
    }
    await new Promise((r) => setTimeout(r, pollIntervalMs));
  }
}

interface DecideApprovalResponseBody {
  decision: "approved" | "denied";
  transaction_id: string | null;
  tx_hash: string | null;
  signature: string | null;
  x_payment_header?: string | null;
  /** Present on the chat-approval-disabled 403 response. */
  error?: string;
  message?: string;
}

async function decide(
  transport: Transport,
  approvalId: string,
  decision: "approved" | "denied",
): Promise<DecideApprovalResult> {
  const { status, body } = await transport.request<DecideApprovalResponseBody>({
    method: "POST",
    path: `/api/approvals/${approvalId}/decide-by-agent`,
    body: { decision },
    expectStatuses: [200, 403],
  });

  if (status === 403 && body.error === "chat_approval_disabled") {
    throw new CanopyChatApprovalDisabledError(approvalId, body.message);
  }

  return {
    decision: body.decision,
    transactionId: body.transaction_id,
    txHash: body.tx_hash,
    signature: body.signature,
  };
}

export async function approve(
  transport: Transport,
  approvalId: string,
): Promise<DecideApprovalResult> {
  return decide(transport, approvalId, "approved");
}

export async function deny(
  transport: Transport,
  approvalId: string,
): Promise<DecideApprovalResult> {
  return decide(transport, approvalId, "denied");
}

File: python/canopy_ai/approval.py
import time
from typing import Literal

from canopy_ai.errors import (
    CanopyApprovalTimeoutError,
    CanopyChatApprovalDisabledError,
)
from canopy_ai.transport import Transport
from canopy_ai.types import ApprovalStatus, DecideApprovalResult


def get_approval_status(transport: Transport, approval_id: str) -> ApprovalStatus:
    _, body = transport.request(
        "GET",
        f"/api/approvals/{approval_id}/status",
        expect_statuses=[200],
    )
    assert isinstance(body, dict)
    return ApprovalStatus(
        status=body["status"],
        decided_at=body.get("decided_at"),
        expires_at=body["expires_at"],
        transaction_id=body["transaction_id"],
        x_payment_header=body.get("x_payment_header"),
    )


def wait_for_approval(
    transport: Transport,
    approval_id: str,
    *,
    timeout_ms: int = 5 * 60 * 1000,
    poll_interval_ms: int = 2000,
) -> ApprovalStatus:
    deadline = time.monotonic() + (timeout_ms / 1000)
    while True:
        status = get_approval_status(transport, approval_id)
        if status["status"] != "pending":
            return status
        if time.monotonic() >= deadline:
            raise CanopyApprovalTimeoutError(approval_id, timeout_ms)
        time.sleep(poll_interval_ms / 1000)


def _decide(
    transport: Transport,
    approval_id: str,
    decision: Literal["approved", "denied"],
) -> DecideApprovalResult:
    status, body = transport.request(
        "POST",
        f"/api/approvals/{approval_id}/decide-by-agent",
        json={"decision": decision},
        expect_statuses=[200, 403],
    )
    assert isinstance(body, dict)
    if status == 403 and body.get("error") == "chat_approval_disabled":
        raise CanopyChatApprovalDisabledError(approval_id, body.get("message"))
    return DecideApprovalResult(
        decision=body["decision"],
        transaction_id=body.get("transaction_id"),
        tx_hash=body.get("tx_hash"),
        signature=body.get("signature"),
    )


def approve(transport: Transport, approval_id: str) -> DecideApprovalResult:
    return _decide(transport, approval_id, "approved")


def deny(transport: Transport, approval_id: str) -> DecideApprovalResult:
    return _decide(transport, approval_id, "denied")
